### PR TITLE
Remove unnecessary variable length arrays.

### PR DESCRIPTION
**Pull request by robertsipka**
VLAs cannot be initialized by any form of initialization syntax with C99 standard.

JerryScript-DCO-1.0-Signed-off-by: Robert Sipka rsipka.uszeged@partner.samsung.com

### CODE REVIEW

**Comment by akiss77** [comment]
The `ECMA_STRING_CONTAINER_UINT32_IN_DESC` paths in `ecma_string_get_size` and `ecma_string_get_length` are very close clones of each other. The first one has no VLA actually.

What about factoring the clones out into a helper function and using:
```c
const uint32_t nums_with_ascending_length[] = { ... };
const uint32_t max_uint32_len = sizeof(nums_with_ascending_length) / sizeof(uint32_t);
```
This would avoid multiplication.

BTW, while we are at the refactoring of these functions: the switch/case in `ecma_string_get_size` looks much better to me than the if/else if/else if/else chain in `ecma_string_get_length`.

BTW2, the implementation of the `ECMA_STRING_CONTAINER_HEAP_NUMBER` paths is also suspiciously similar in the two functions. Candidates for yet another helper?

**Comment by robertsipka** [comment]
Thanks the comments, I've updated this PR.

File: jerry-core/ecma/base/ecma-helpers-string.cpp [hunk 1]
```diff
@@ -1286,6 +1286,42 @@ ecma_compare_ecma_strings_relational (const ecma_string_t *string1_p, /**< ecma-
 } /* ecma_compare_ecma_strings_relational */
 
 /**
+ * Lengths for numeric string values
+ */
+const uint32_t nums_with_ascending_length[] =
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.cpp]: Should be `static`

File: jerry-core/ecma/base/ecma-helpers-string.cpp [hunk 2]
```diff
@@ -1286,6 +1286,42 @@ ecma_compare_ecma_strings_relational (const ecma_string_t *string1_p, /**< ecma-
...
+  1000000u,
+  10000000u,
+  100000000u,
+  1000000000u
+};
+
+/**
+ * Maximum length of numeric strings
+ */
+const uint32_t max_uint32_len = sizeof (nums_with_ascending_length) / sizeof (uint32_t);
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.cpp]: ditto

File: jerry-core/ecma/base/ecma-helpers-string.cpp [hunk 3]
```diff
@@ -1293,75 +1329,55 @@ ecma_compare_ecma_strings_relational (const ecma_string_t *string1_p, /**< ecma-
...
+      {
+        length++;
+      }
 
-    return (ecma_length_t) ecma_number_to_utf8_string (*num_p, buffer, sizeof (buffer));
-  }
-  else
-  {
-    JERRY_ASSERT (container == ECMA_STRING_CONTAINER_HEAP_CHUNKS);
+      return length;
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.cpp]: This `ECMA_STRING_CONTAINER_UINT32_IN_DESC` case is still here, a clone of the same case in `ecma_string_get_size`, and not factored out into a helper. Intentional?

**Comment by robertsipka** [comment]
Thanks again, I've fixed this based on your suggestions.

File: jerry-core/ecma/base/ecma-helpers-string.cpp [hunk 4]
```diff
@@ -1286,6 +1286,64 @@ ecma_compare_ecma_strings_relational (const ecma_string_t *string1_p, /**< ecma-
...
+
+/**
+ * Get size of the number stored locally in the string's descriptor
+ *
+ * Note: the represented number size and length are equal
+ *
+ * @return size in bytes
+ */
+static ecma_length_t
+ecma_string_get_number_in_desc_size (const uint32_t uint32_number)
```
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.cpp]: Please add comment on argument.
moreover, I'd swap the order of these two new helper functions to get this one closer to the data it uses (`nums_with_ascending_length` and `max_uint32_len`).

**Comment by robertsipka** [comment]
Fixed.

**Comment by akiss77** [comment]
LGTM (informally)

**Comment by LaszloLango** [comment]
@robertsipka, Could you measure this patch on RPi2 or at least on your own PC? @bzsolt, can you help with the measurement on RPi2? We have to be careful when refactoring these functions, because they are called really often.

**Comment by bzsolt** [comment]
The RPi2 results with run-perf-test.sh
master: [91a0514](https://github.com/Samsung/jerryscript/commit/91a0514fab38ed7c82e1791d5ff9813a889ddaea)

Benchmark | RSS<br>(+ is better) | Perf<br>(+ is better)
---------: | --------- | ---------
3d-cube.js | `  96k ->   92k      `<br>`             +4.167%` | `  2.700s ->   2.700s                      `<br>`                 +0.000%  (+-0.862%) : [≈]`
access-binary-trees.js | `  84k ->   84k      `<br>`             +0.000%` | `  1.490s ->   1.505s                      `<br>`                 -1.007%  (+-1.167%) : [≈]`
access-fannkuch.js | `  40k ->   40k      `<br>`             +0.000%` | `  8.035s ->   8.177s                      `<br>`                 -1.763%  (+-0.805%) : [-]`
access-nbody.js | `  48k ->   44k      `<br>`             +8.333%` | `  3.230s ->   3.357s                      `<br>`                 -3.922%  (+-0.962%) : [-]`
bitops-3bit-bits-in-byte.js | `  32k ->   32k      `<br>`             +0.000%` | `  1.762s ->   1.797s                      `<br>`                 -1.987%  (+-1.054%) : [-]`
bitops-bits-in-byte.js | `  32k ->   32k      `<br>`             +0.000%` | `  2.690s ->   2.712s                      `<br>`                 -0.806%  (+-0.818%) : [≈]`
bitops-bitwise-and.js | `  32k ->   32k      `<br>`             +0.000%` | `  3.728s ->   3.805s                      `<br>`                 -2.056%  (+-0.646%) : [-]`
controlflow-recursive.js | ` 152k ->  148k      `<br>`             +2.632%` | `  0.900s ->   0.920s                      `<br>`                 -2.222%  (+-0.000%) : [-]`
crypto-aes.js | ` 124k ->  124k      `<br>`             +0.000%` | `  4.752s ->   4.792s                      `<br>`                 -0.842%  (+-1.097%) : [≈]`
crypto-md5.js | ` 184k ->  184k      `<br>`             +0.000%` | ` 21.153s ->  21.138s                      `<br>`                 +0.071%  (+-0.748%) : [≈]`
crypto-sha1.js | ` 132k ->  128k      `<br>`             +3.030%` | `  9.582s ->   9.637s                      `<br>`                 -0.574%  (+-0.714%) : [≈]`
date-format-tofte.js | `  64k ->   64k      `<br>`             +0.000%` | `  2.225s ->   2.233s                      `<br>`                 -0.374%  (+-0.558%) : [≈]`
date-format-xparb.js | `  64k ->   64k      `<br>`             +0.000%` | `  1.145s ->   1.145s                      `<br>`                 +0.000%  (+-1.114%) : [≈]`
math-cordic.js | `  32k ->   32k      `<br>`             +0.000%` | `  3.040s ->   3.150s                      `<br>`                 -3.618%  (+-0.783%) : [-]`
math-partial-sums.js | `  36k ->   36k      `<br>`             +0.000%` | `  1.853s ->   1.910s                      `<br>`                 -3.058%  (+-1.082%) : [-]`
math-spectral-norm.js | `  44k ->   44k      `<br>`             +0.000%` | `  1.535s ->   1.547s                      `<br>`                 -0.760%  (+-0.811%) : [≈]`
string-fasta.js | `  48k ->   48k      `<br>`             +0.000%` | `  4.607s ->   4.620s                      `<br>`                 -0.289%  (+-0.584%) : [≈]`
Geometric mean: | RSS reduction: `1.094%` | Speed up: `-1.358% (+-0.205%) : [-]`

**Comment by LaszloLango** [comment]
Could you try how the performance change if you make the `ecma_string_get_number_in_desc_size` and `ecma_string_get_heap_number_size` inline with `__attr_always_inline___`?

**Comment by bzsolt** [comment]
Of course, the results are here:

Benchmark | RSS<br>(+ is better) | Perf<br>(+ is better)
---------: | --------- | ---------
3d-cube.js | `  96k ->   96k      `<br>`             +0.000%` | `  2.702s ->   2.710s                      `<br>`                 -0.308%  (+-0.898%) : [≈]`
access-binary-trees.js | `  84k ->   84k      `<br>`             +0.000%` | `  1.487s ->   1.495s                      `<br>`                 -0.560%  (+-0.836%) : [≈]`
access-fannkuch.js | `  40k ->   40k      `<br>`             +0.000%` | `  8.035s ->   8.187s                      `<br>`                 -1.888%  (+-0.560%) : [-]`
access-nbody.js | `  48k ->   44k      `<br>`             +8.333%` | `  3.238s ->   3.270s                      `<br>`                 -0.978%  (+-1.133%) : [≈]`
bitops-3bit-bits-in-byte.js | `  32k ->   32k      `<br>`             +0.000%` | `  1.755s ->   1.750s                      `<br>`                 +0.285%  (+-0.783%) : [≈]`
bitops-bits-in-byte.js | `  32k ->   32k      `<br>`             +0.000%` | `  2.690s ->   2.677s                      `<br>`                 +0.495%  (+-0.738%) : [≈]`
bitops-bitwise-and.js | `  32k ->   32k      `<br>`             +0.000%` | `  3.713s ->   3.823s                      `<br>`                 -2.962%  (+-0.590%) : [-]`
controlflow-recursive.js | ` 152k ->  148k      `<br>`             +2.632%` | `  0.898s ->   0.905s                      `<br>`                 -0.742%  (+-1.255%) : [≈]`
crypto-aes.js | ` 124k ->  124k      `<br>`             +0.000%` | `  4.753s ->   4.765s                      `<br>`                 -0.246%  (+-0.673%) : [≈]`
crypto-md5.js | ` 184k ->  184k      `<br>`             +0.000%` | ` 21.153s ->  21.085s                      `<br>`                 +0.323%  (+-0.799%) : [≈]`
crypto-sha1.js | ` 132k ->  128k      `<br>`             +3.030%` | `  9.585s ->   9.588s                      `<br>`                 -0.035%  (+-0.960%) : [≈]`
date-format-tofte.js | `  64k ->   64k      `<br>`             +0.000%` | `  2.230s ->   2.237s                      `<br>`                 -0.299%  (+-1.294%) : [≈]`
date-format-xparb.js | `  64k ->   64k      `<br>`             +0.000%` | `  1.143s ->   1.145s                      `<br>`                 -0.146%  (+-1.416%) : [≈]`
math-cordic.js | `  32k ->   32k      `<br>`             +0.000%` | `  3.037s ->   3.033s                      `<br>`                 +0.110%  (+-0.791%) : [≈]`
math-partial-sums.js | `  36k ->   36k      `<br>`             +0.000%` | `  1.853s ->   1.863s                      `<br>`                 -0.540%  (+-1.027%) : [≈]`
math-spectral-norm.js | `  44k ->   44k      `<br>`             +0.000%` | `  1.537s ->   1.522s                      `<br>`                 +0.976%  (+-0.701%) : [+]`
string-fasta.js | `  48k ->   48k      `<br>`             +0.000%` | `  4.618s ->   4.625s                      `<br>`                 -0.144%  (+-0.787%) : [≈]`
Geometric mean: | RSS reduction: `0.846%` | Speed up: `-0.388% (+-0.225%) : [-]`

**Comment by LaszloLango** [comment]
@bzsolt, looks better.
@robertsipka, please update the PR.

**Comment by robertsipka** [comment]
@bzsolt @LaszloLango : thanks, fixed.

**Comment by LaszloLango** [comment]
LGTM

**Comment by LaszloLango** [comment]
@robertsipka, precommit failed. Please fix it.

**Comment by robertsipka** [comment]
@LaszloLango : There is no error with gcc 4.9.2. I checked with gcc 4.8.4, and the fail was occurred. I've fixed it, thanks!

**Pull request merged**